crates/holochain_types/src/chain.rs

```diff
@@ -46,6 +48,64 @@ pub struct ChainFilterIter {
     end: bool,
 }
 
+#[derive(Debug, Clone, PartialEq, Eq)]
+/// A [`ChainFilter`] with the action sequences for the
+/// starting position and any `until` hashes.
+pub struct ChainFilterRange {
+    /// The filter for for this chain.
+    filter: ChainFilter,
+    /// The start of this range is the end of
+    /// the filter iterator.
+    /// The end of this range is the sequence of
+    /// the starting position hash.
+    range: RangeInclusive<u32>,
+    /// The start of the ranges type.
+    range_start_type: RangeStartType,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+/// The type of chain item that starts the range.
+enum RangeStartType {
+    /// The range starts from genesis.
+    Genesis,
+    /// The Range starts from an action where `take`
+    /// has reached zero.
+    Take,
+    /// The range starts from an action where an
+    /// `until` hash was found.
+    Until,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+/// Outcome of trying to find the action sequences in a filter.
+pub enum Sequences {
+    /// Found all action sequences
+    Found(ChainFilterRange),
+    /// The following action was not found.
+    ActionNotFound(ActionHash),
```

Comment:
Name suggestion: `PositionNotFound`, since this only gets triggered if the position is not found, not if an Until isn't found, right? 

But I also think it could be good to rename "position" to "top", so maybe `FilterTopNotFound`?